pasha/functor.py

```diff
@@ -0,0 +1,280 @@
+    classmethod, then it can take advantage of automatic wrapping of
+    values passed to a map call.
+
+    Alternatively, a type implement an '_pasha_functor_ method to
+    construct a sletabjn functor. This will always take precedence over
+    the automatic detection.
+    """
+
+    _functor_types = []
+
+    @classmethod
+    def __init_subclass__(cls):
+        cls._functor_types.append(cls)
+
+    @classmethod
+    def try_wrap(cls, value):
+        """Attempt to wrap a value in a functor.
+
+        Args:
+            value (Any): Value to wrap for map operation.
+
+        Returns:
+            (Functor) Functor wrapping the given value or the same value
+                if already a subtype of Functor.
+
+        Raises:
+            ValueError: If no or more than one default functor types
+                could be applied.
+        """
+
+        if isinstance(value, cls):
+            return value
+
+        if hasattr(value, '_pasha_functor_'):
+            return value._pasha_functor_()
+
+        functor = None
+
+        for functor_type in cls._functor_types:
+            cur_functor = functor_type.wrap(value)
+
+            if cur_functor is not None:
+                if functor is not None:
+                    raise ValueError('default functor is ambiguous')
+
+                functor = cur_functor
+
+        if functor is None:
+            raise ValueError(f'no default functor for {type(value)}')
+
+        return functor
+
+    @classmethod
+    def wrap(self, value):
+        """Wrap value in this functor type, if possible.
+
+        Args:
+            value (Any): Value to wrap for map operation.
+
+        Returns:
+            (Functor) Functor if wrapping is possible or None.
+        """
+
+        return
+
+    def split(self, num_workers):
+        """Split this functor into work units.
+
+        The values contained in the returned Iterable are passed to this
+        functor's iterate method later on. It may consist of any value
+        suitable to describe each work unit, e.g. an iterable of indices
+        of a sequence.
+
+        Args:
+            num_workers (int): Number of workers processing this
+                functor.
+
+        Returns:
+            (Iterable) Iterable of elements for each work unit.
+        """
+
+        raise NotImplementedError('split')
+
+    def iterate(self, share):
+        """Iterate over a share of this functor.
+
+        Args:
+            share (Any): Element of the Iterable returned by split() to
+                iterate over. The exact type of specification is up to
+                the `Functor` implementation.
+
+        Returns:
+            None
+        """
+
+        raise NotImplementedError('iterate')
+
+
+class SequenceFunctor(Functor):
+    """Functor wrapping a generic sequence.
+
+    This functor can wrap any indexable collection, e.g. list, tuples,
+    or any other type implementing __getitem__. It automatically wraps
```

Comment:
Nitpick: `__getitem__` is necessary but not sufficient for a sequence. Most obviously, mappings have `__getitem__` but won't work with this.